Introduce vFloat and float types in ParameterSets

felicepantaleo opened the pull request:
### Description

This pull request introduces `vFloat` and `float` types in the `ParameterSets` of the CMSSW framework. The changes aim to improve the precision and performance of floating-point parameter handling in configuration sets.

A new Entry type code is added for float/vfloat. Files that store a float parameter in their configuration provenance cannot be read by releases that do not contain this change (FormatIncompatibility: Unknown ParameterSet Entry type encoding: 'H'). Encodings of existing double parameters are unchanged.

### Changes

1. Added `vFloat` and `float` type definitions.
2. Updated relevant configuration files to utilize the new types.
3. Ensured compatibility with existing functionality and modules.

This will allow the developers of C++ modules to directly use `std::vector<float>` configuration parameters avoiding manually converting each of them from  `std::vector<double>`, if double precision is not needed.


### Testing

- Ran all existing unit tests.
- Added new tests to cover the `vFloat` and `float` types.
- Tested with workflow `24900.0`.
- Verified no regressions in performance or functionality.

### Future Work

If the tests run smoothly, an existing reconstruction module will be migrated from `double` to `float` within this pull request.

@smuzaffar @makortel @Dr15Jones @fwyzard @rovere @VinInn FYI

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"000b96e13c45cc7b28cc549ba777f96b2f22254a":{"files":["FWCore/ParameterSet/src/types.cc","FWCore/ParameterSet/test/test_catch2_ps.cc"],"squashed":true,"time":1785201313},"1595eb97ad8720a94bc2051dba42b0008e222934":{"files":["FWCore/Integration/plugins/ProducerWithPSetDesc.cc","FWCore/ParameterSet/src/types.cc","FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc","FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc","FWCore/ParameterSet/src/ParameterDescription.cc"],"squashed":false,"time":1785295014},"1605f41f71479f4202f0443c5efaf5e8b48b565c":{"files":["FWCore/Integration/plugins/ProducerWithPSetDesc.cc","FWCore/ParameterSet/src/types.cc","FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc","FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc","FWCore/ParameterSet/src/ParameterDescription.cc"],"squashed":false,"time":1785294999},"19e8bb180315e22e1652466664f9c1b96f07c675":{"files":["FWCore/ParameterSet/python/Config.py","FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_briefdoc.txt","FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_doc.txt","FWCore/ParameterSet/python/Types.py","FWCore/Integration/plugins/ProducerWithPSetDesc.cc","FWCore/ParameterSet/test/test_catch2_ps.cc","FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_cfi.py","FWCore/ParameterSet/src/ParameterDescription.cc"],"squashed":true,"time":1785049364},"2ce55515946653e1b7f1310c5933e7902808e0c6":{"files":["FWCore/ParameterSet/interface/Entry.h","FWCore/ParameterSet/interface/ParameterDescription.h","FWCore/ParameterSet/interface/ParameterDescriptionNode.h","FWCore/ParameterSet/interface/ParameterSet.h","FWCore/ParameterSet/interface/types.h","FWCore/ParameterSet/python/Config.py","FWCore/ParameterSet/python/DummyCfis.py","FWCore/ParameterSet/python/Types.py","FWCore/ParameterSet/src/Entry.cc","FWCore/ParameterSet/src/ParameterDescription.cc","FWCore/ParameterSet/src/ParameterDescriptionNode.cc","FWCore/ParameterSet/src/ParameterSet.cc","FWCore/ParameterSet/src/types.cc","FWCore/ParameterSet/test/cmsconfig.py","FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc","FWCore/PythonParameterSet/src/PyBind11Module.h","IOMC/ParticleGuns/interface/CloseByParticleGunProducer.h","IOMC/ParticleGuns/src/CloseByParticleGunProducer.cc"],"squashed":true,"time":1718187576},"4e3d218223807897f2771457d06ffc62d642c5ba":{"files":["FWCore/ParameterSet/interface/Entry.h","FWCore/ParameterSet/interface/ParameterDescription.h","FWCore/ParameterSet/interface/ParameterDescriptionNode.h","FWCore/ParameterSet/interface/ParameterSet.h","FWCore/ParameterSet/interface/types.h","FWCore/ParameterSet/python/Config.py","FWCore/ParameterSet/python/DummyCfis.py","FWCore/ParameterSet/python/Types.py","FWCore/ParameterSet/src/Entry.cc","FWCore/ParameterSet/src/ParameterDescription.cc","FWCore/ParameterSet/src/ParameterDescriptionNode.cc","FWCore/ParameterSet/src/ParameterSet.cc","FWCore/ParameterSet/src/types.cc","FWCore/ParameterSet/test/cmsconfig.py","FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc","FWCore/PythonParameterSet/src/PyBind11Module.h"],"squashed":true,"time":1718275866},"591ad8e7b87019be98faef9cd43bcee5b16af857":{"files":["RecoLocalTracker/SiStripRecHitConverter/src/StripCPEfromTrackAngle.cc"],"squashed":true,"time":1726476575},"618c90ba08e0038888b2cfd45776f76152e5dd57":{"files":["FWCore/Integration/plugins/ProducerWithPSetDesc.cc","FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_briefdoc.txt","FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_cfi.py","FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_doc.txt","FWCore/ParameterSet/scripts/edmPythonConfigToCppValidation","FWCore/ParameterSet/src/FillDescriptionFromPSet.cc","FWCore/ParameterSet/src/ParameterDescriptionBase.cc"],"squashed":true,"time":1785201116},"6d0468bc0f0dd92d9ade1e2350b94298a2b6184c":{"files":["FWCore/ParameterSet/src/ParameterDescriptionNode.cc","FWCore/ParameterSet/interface/Entry.h","FWCore/ParameterSet/python/Types.py","FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc","FWCore/ParameterSet/interface/ParameterDescription.h","RecoLocalTracker/SiStripRecHitConverter/src/StripCPEfromTrackAngle.cc","FWCore/ParameterSet/test/cmsconfig.py","FWCore/ParameterSet/interface/ParameterDescriptionNode.h","FWCore/ParameterSet/src/ParameterDescription.cc","FWCore/ParameterSet/src/Entry.cc","FWCore/ParameterSet/python/Config.py"],"squashed":true,"time":1783239280},"6fdc352609e26b9f027cfd96e2ed1630324e7d2e":{"files":["FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc"],"squashed":true,"time":1785202454},"731e8ada85b7eeaecc5afa36b0b08079139e2cb1":{"files":["FWCore/Integration/plugins/ProducerWithPSetDesc.cc","FWCore/ParameterSet/src/types.cc","FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc","FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc","FWCore/ParameterSet/src/ParameterDescription.cc"],"squashed":false,"time":1785295013},"7e6433d1fe45fa891f2ae22ab7b0ac878644cad6":{"files":["FWCore/ParameterSet/interface/Entry.h","FWCore/ParameterSet/interface/ParameterDescription.h","FWCore/ParameterSet/interface/ParameterDescriptionNode.h","FWCore/ParameterSet/interface/ParameterSet.h","FWCore/ParameterSet/interface/types.h","FWCore/ParameterSet/python/Config.py","FWCore/ParameterSet/python/DummyCfis.py","FWCore/ParameterSet/python/Types.py","FWCore/ParameterSet/src/Entry.cc","FWCore/ParameterSet/src/ParameterDescription.cc","FWCore/ParameterSet/src/ParameterDescriptionNode.cc","FWCore/ParameterSet/src/ParameterSet.cc","FWCore/ParameterSet/src/types.cc","FWCore/ParameterSet/test/cmsconfig.py","FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc","FWCore/PythonParameterSet/src/PyBind11Module.h"],"squashed":true,"time":1725884157},"8a097d70ceefedd30290d354179e0e0d42b5bf42":{"files":["FWCore/Integration/plugins/ProducerWithPSetDesc.cc","FWCore/ParameterSet/src/types.cc","FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc","FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc","FWCore/ParameterSet/src/ParameterDescription.cc"],"squashed":false,"time":1785295013},"98976752ac57e575c1a85597e602da966c97a1f2":{"files":["FWCore/Integration/plugins/ProducerWithPSetDesc.cc","FWCore/ParameterSet/src/types.cc","FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc","FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc","FWCore/ParameterSet/src/ParameterDescription.cc"],"squashed":false,"time":1785295028},"9f53fdd07b29a990689fdc01d0d0ebcd53945804":{"files":["FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc"],"squashed":true,"time":1785202835},"af69b627d71c767de41f633fb3a38fa3675e2127":{"files":["FWCore/ParameterSet/python/Config.py","FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_briefdoc.txt","FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_doc.txt","FWCore/ParameterSet/python/Types.py","FWCore/Integration/plugins/ProducerWithPSetDesc.cc","FWCore/ParameterSet/test/test_catch2_ps.cc","FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_cfi.py","FWCore/ParameterSet/src/ParameterDescription.cc"],"squashed":true,"time":1785049364},"c6877d88b39f17575f322ed72406378a564f33a7":{"files":["FWCore/ParameterSet/python/Config.py","FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_briefdoc.txt","FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_doc.txt","FWCore/ParameterSet/python/Types.py","FWCore/Integration/plugins/ProducerWithPSetDesc.cc","FWCore/ParameterSet/test/test_catch2_ps.cc","FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_cfi.py","FWCore/ParameterSet/src/ParameterDescription.cc"],"squashed":true,"time":1785048608},"c97cd2912d52b8d16e5146fde8b298090dcc27df":{"files":["FWCore/ParameterSet/src/Entry.cc"],"squashed":true,"time":1726484102},"cff29d8de9ce4c9339431f2833f7c72e366512ee":{"files":["FWCore/Integration/plugins/ProducerWithPSetDesc.cc","FWCore/ParameterSet/src/types.cc","FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc","FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc","FWCore/ParameterSet/src/ParameterDescription.cc"],"squashed":false,"time":1785295013},"d9c1ac770109e6f1468e8970273198b973708684":{"files":["FWCore/ParameterSet/python/Config.py","FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_briefdoc.txt","FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_doc.txt","FWCore/ParameterSet/python/Types.py","FWCore/Integration/plugins/ProducerWithPSetDesc.cc","FWCore/ParameterSet/test/test_catch2_ps.cc","FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_cfi.py","FWCore/ParameterSet/src/ParameterDescription.cc"],"squashed":true,"time":1785048105},"f09ae150b991de473848b1b5c9be542130b4b504":{"files":["FWCore/Integration/plugins/ProducerWithPSetDesc.cc","FWCore/ParameterSet/src/types.cc","FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc","FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc","FWCore/ParameterSet/src/ParameterDescription.cc"],"squashed":false,"time":1785295028}},"emoji":{"2165589181":"+1","2165589927":"+1","2166817034":"+1","2166818611":"+1","2345608764":"+1","2352858070":"+1","2355252253":"+1","2550593394":"+1","4885672708":"+1","5083004531":"+1","5099535548":"+1","5099687671":"+1","5103532196":"+1","5113592100":"+1","5147145142":"+1","5156205690":"+1"},"last_seen_sha":"98976752ac57e575c1a85597e602da966c97a1f2","signatures":{"2355252253":"c97cd2912d52b8d16e5146fde8b298090dcc27df","5147145142":"f09ae150b991de473848b1b5c9be542130b4b504","5156205690":"f09ae150b991de473848b1b5c9be542130b4b504"}} -->

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45216/40573

- This PR adds an extra 132KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #43276

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45216/40573/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45216/40573/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

felicepantaleo committed:
Introduce vFloat and float types in ParameterSets

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45216/40574

- This PR adds an extra 124KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #43276

cmsbuild commented:
A new Pull Request was created by  @felicepantaleo for master.

It involves the following packages:

- FWCore/ParameterSet (**core**)
- FWCore/PythonParameterSet (**core**)


@cmsbuild, @makortel, @Dr15Jones, @smuzaffar can you please review it and eventually sign? Thanks.
@makortel, @missirol, @wddgit this is something you requested to watch as well.
@sextonkennedy, @rappoccio, @antoniovilela you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

felicepantaleo commented:
test parameters:

-  workflow = 25288.203,25488.203

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
-1

**Failed Tests**: RelVals 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9ee9da/39871/summary.html
**COMMIT**: 4e3d218223807897f2771457d06ffc62d642c5ba
**CMSSW**: CMSSW_14_1_X_2024-06-13-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/45216/39871/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals

ValueError: Undefined workflows: 25288.203, 25488.203

felicepantaleo commented:
test parameters:

felicepantaleo commented:
@cmsbuild please test

Dr15Jones commented:
Given @makortel is on vacation until beginning of July, I do not plan on doing a deep dive on the PR until he gets back.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9ee9da/39878/summary.html
**COMMIT**: 4e3d218223807897f2771457d06ffc62d642c5ba
**CMSSW**: CMSSW_14_1_X_2024-06-13-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/45216/39878/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 108 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-06-13-1100+9ee9da/63184/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 45263 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-06-13-1100+9ee9da/63184/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3345018
 - DQMHistoTests: [Total failures: 27593](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-06-13-1100+9ee9da/63184/validateJR.html)
 - DQMHistoTests: Total nulls: 24
 - DQMHistoTests: Total successes: 3317381
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.013000000000000001 KiB( 47 files compared)
 - DQMHistoSizes: changed ( 1000.0 ): 0.008 KiB        JetMET/SUSYDQM
 - DQMHistoSizes: changed ( 140.023 ): -0.016 KiB       JetMET/SUSYDQM
 - DQMHistoSizes: changed ( 140.043 ): -0.008 KiB       JetMET/SUSYDQM
 - DQMHistoSizes: changed ( 140.063 ): 0.023 KiB        JetMET/SUSYDQM
 - DQMHistoSizes: changed ( 141.042 ): -0.020 KiB       JetMET/SUSYDQM
 - DQMHistoSizes: changed ( 141.044 ): 0.016 KiB        JetMET/SUSYDQM
 - DQMHistoSizes: changed ( 141.046 ): -0.012 KiB       JetMET/SUSYDQM
 - DQMHistoSizes: changed ( 4.53 ): -0.004 KiB       JetMET/SUSYDQM
 - Checked 202 log files, 165 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-06-13-1100+9ee9da/63184/triggerResults): found differences in 10 / 46 workflows

felicepantaleo commented:
no worries @Dr15Jones , there is still a bit of hammering to do before Matti comes back

felicepantaleo commented:
I have checked the changes in the DQM and they are due to the ~2000 floating point parameters in modules that are not using the `fillDescriptions` and are not declared as `cms.double`. These are now using `cms.float` by default. I will check now how to make `cms.double` the default for these.

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_14_2_X. Please open a backport if it should also go in to CMSSW_14_1_X.

antoniovilela commented:
ping (to make bot change milestone)

felicepantaleo commented:
@cmsbuild please test

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45216/41803

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #43276

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45216/41803/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45216/41803/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45216/41805

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #43276

cmsbuild commented:
Pull request #45216 was updated. @Dr15Jones, @cmsbuild, @jfernan2, @makortel, @mandrenguyen, @smuzaffar can you please check and sign again.

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 28KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9ee9da/41534/summary.html
**COMMIT**: c97cd2912d52b8d16e5146fde8b298090dcc27df
**CMSSW**: CMSSW_14_2_X_2024-09-15-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/45216/41534/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 3 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-15-2300+9ee9da/64557/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 5 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-15-2300+9ee9da/64557/validateJR.html)
 - DQMHistoTests: Total files compared: 44
 - DQMHistoTests: Total histograms compared: 3331158
 - DQMHistoTests: [Total failures: 6](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-15-2300+9ee9da/64557/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3331132
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 43 files compared)
 - Checked 193 log files, 163 edm output root files, 44 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-15-2300+9ee9da/64557/triggerResults): no differences found

jfernan2 commented:
+1

makortel commented:
> This will allow the developers of C++ modules to directly use `std::vector<float>` configuration parameters avoiding manually converting each of them from `std::vector<double>`, if double precision is not needed.

There are simpler ways to achieve an easier conversion from `std::vector<double>` to `std::vector<float>`, or enabling `getParameter<std::vector<float>>(...)`, that would not require changes in the python configuration side. Could you describe what aspects, in addition to what you wrote in the PR description, you see as important so that it floating point precision is useful to be specified in the python configuration side as well?

(based on a cursory internal discussion we are leaning towards a separate configuration data type to be a better way, e.g. for consistency with the integer types and requiring consistency between `fillDescriptions()` and `getParameter()`, so this question is only about us learning what additional motivations do you have in mind)

felicepantaleo commented:
@makortel let me know if you have more comments

makortel commented:
@felicepantaleo I'll have some comments, but have been distracted by other stuff (and will be for some days).

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_15_0_X. Please open a backport if it should also go in to CMSSW_14_2_X.

smuzaffar commented:
code-checks

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45216/43099

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #43276

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_15_1_X. Please open a backport if it should also go in to CMSSW_15_0_X.

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_16_1_X. Please open a backport if it should also go in to CMSSW_16_0_X.

cmsbuild commented:
-1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9ee9da/50336/summary.html
**COMMIT**: c97cd2912d52b8d16e5146fde8b298090dcc27df
**CMSSW**: CMSSW_16_0_X_2025-12-17-2300/el8_amd64_gcc13
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/45216/50336/install.sh` to create a dev area with all the needed externals and cmssw changes.

This pull request cannot be automatically merged, could you please rebase it?
You can see the log for git cms-merge-topic here: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9ee9da/50336/git-merge-result

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_20_0_X. Please open a backport if it should also go in to CMSSW_17_0_X.

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_20_1_X. Please open a backport if it should also go in to CMSSW_20_0_X.

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45216/50078

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #51389
  - File  FWCore/ParameterSet/src/Entry.cc  modified in PR(s): #50585

cmsbuild commented:
Pull request #45216 was updated. @Dr15Jones, @makortel, @smuzaffar can you please check and sign again.

cmsbuild commented:
+1

**Size**: This PR adds an extra 104KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9ee9da/54525/summary.html
**COMMIT**: 6d0468bc0f0dd92d9ade1e2350b94298a2b6184c
**CMSSW**: CMSSW_20_1_X_2026-07-05-0000/el9_amd64_gcc13
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/45216/54525/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_20_1_X_2026-07-05-0000+9ee9da/76694/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_20_1_X_2026-07-05-0000+9ee9da/76694/validateJR.html)
 - DQMHistoTests: Total files compared: 45
 - DQMHistoTests: Total histograms compared: 3547841
 - DQMHistoTests: [Total failures: 16](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_20_1_X_2026-07-05-0000+9ee9da/76694/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3547807
 - DQMHistoTests: Total skipped: 18
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 44 files compared)
 - Checked 195 log files, 163 edm output root files, 45 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_20_1_X_2026-07-05-0000+9ee9da/76694/triggerResults): no differences found

Martin-Grunewald commented:
Can you confirm that this is only targetting CMSSW_2X_Y / Phase-2, ie, no backport to 17_0 / Run-3?
Otherwise we'd eventually need to update the (old/Run-3) ConfDb to handle these additional types, which we want to avoid.

makortel commented:
I think this PR should target only 20_1_X.

felicepantaleo commented:
> Can you confirm that this is only targetting CMSSW_2X_Y / Phase-2, ie, no backport to 17_0 / Run-3? Otherwise we'd eventually need to update the (old/Run-3) ConfDb to handle these additional types, which we want to avoid.

yes, only CMSSW_20_1_X. No backport is foreseen.

makortel commented:
> This will allow the developers of C++ modules to directly use `std::vector<float>` configuration parameters avoiding manually converting each of them from `std::vector<double>`, if double precision is not needed.

Just to note that allowing `vector<double>` parameter to be obtained as `vector<float>` from the `edm::ParameterSet` would be fairly straightforward with (IIRC) less than 10 lines of code (plus tests).

Written that, the distinct single and double precision floating point types in the configuration mirror the handling of the integer types (32 and 64 bits, both as signed and unsigned) and therefore "why not". In a way it would be nice to do some validity checks (e.g. overflows or underflows) already in python, but we don't do those for the integer types either, so I won't ask.

After a first look (I need to do another more detailed one) the following tests would need to be extended

https://github.com/cms-sw/cmssw/blob/6d0468bc0f0dd92d9ade1e2350b94298a2b6184c/FWCore/ParameterSet/python/Types.py#L1803-L1828

----

https://github.com/cms-sw/cmssw/blob/6d0468bc0f0dd92d9ade1e2350b94298a2b6184c/FWCore/ParameterSet/test/test_catch2_ps.cc#L152-L165
https://github.com/cms-sw/cmssw/blob/6d0468bc0f0dd92d9ade1e2350b94298a2b6184c/FWCore/ParameterSet/test/test_catch2_ps.cc#L491-L519
https://github.com/cms-sw/cmssw/blob/6d0468bc0f0dd92d9ade1e2350b94298a2b6184c/FWCore/ParameterSet/test/test_catch2_ps.cc#L646-L647
for testing that one gets the same value back from the `ParameterSet` as the parameter was set, and testing various corner cases (the currently extended tests in `test_catch_ParameterSetDescription.cc` are really about the ParameterSet validation system). I'm puzzled though why the infinity/NaN would not propagate through the ParameterSet, and I'm going to look into that.

----

https://github.com/cms-sw/cmssw/blob/6d0468bc0f0dd92d9ade1e2350b94298a2b6184c/FWCore/Integration/plugins/ProducerWithPSetDesc.cc#L198-L223
Although here I'd nowadays expect the `std::numeric_limits<double>::min()` because the `edm::encode()` uses `std::numeric_limits<double>::max_digits10`. I'll check that as well. 

https://github.com/cms-sw/cmssw/blob/6d0468bc0f0dd92d9ade1e2350b94298a2b6184c/FWCore/Integration/plugins/ProducerWithPSetDesc.cc#L640-L656
Here I wonder why `std::numeric_limits<double>::max()` would fail, so I'll check that.

https://github.com/cms-sw/cmssw/blob/6d0468bc0f0dd92d9ade1e2350b94298a2b6184c/FWCore/Integration/plugins/ProducerWithPSetDesc.cc#L1070-L1071
https://github.com/cms-sw/cmssw/blob/6d0468bc0f0dd92d9ade1e2350b94298a2b6184c/FWCore/Integration/plugins/ProducerWithPSetDesc.cc#L1099-L1100
https://github.com/cms-sw/cmssw/blob/6d0468bc0f0dd92d9ade1e2350b94298a2b6184c/FWCore/Integration/plugins/ProducerWithPSetDesc.cc#L1187-L1188
https://github.com/cms-sw/cmssw/blob/6d0468bc0f0dd92d9ade1e2350b94298a2b6184c/FWCore/Integration/plugins/ProducerWithPSetDesc.cc#L1216-L1217

Extending `ProducerWithPSetDesc` will likely have knock-on effects on the tests that make use of it.

----

I didn't find a systematic test that the ParameterSet stored in a file is read properly e.g. with `edmProvDump`. I'm thinking to add such a test making use of the `ProducerWithPSetDesc`.

felicepantaleo committed:
FWCore/ParameterSet: add float/vfloat value round-trip and corner-case tests

Adds float and vfloat set/get round-trip tests to test_catch2_ps.cc (special
values, limits, equality, negative zero, distinctness from double) and
float/vfloat tests to the Types.py unit tests.

Infinity and NaN do propagate through the ParameterSet for both double and
float; the previously commented-out cases could not use the equality-based
helper because NaN never compares equal to itself, so a dedicated helper is
used. Subnormal values do not survive the round-trip because std::stod/stof
report ERANGE for a subnormal result.

felicepantaleo committed:
FWCore/Integration: exercise float and vfloat in ProducerWithPSetDesc

Adds float and vfloat parameters, their value checks, and noDefault entries to
the ParameterSetDescription integration test, with the regenerated reference
files.

felicepantaleo committed:
FWCore/ParameterSet: write float parameters to cfi with full precision

writeSingleValue<float> fell through to the generic template, which uses the
default stream precision of 6 significant digits, so a float parameter written
into a cfi did not read back as the same value (for example
std::numeric_limits<float>::min() was written as 1.17549e-38). Add formatFloat
and the corresponding specialization, mirroring the existing double handling:
use max_digits10 and fall back to the shorter digits10 form only when it
converts back to the same value.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45216/50311

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/ParameterSet/interface/Entry.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/ParameterDescription.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/ParameterDescriptionNode.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/ParameterSet.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/types.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #51413
  - File  FWCore/ParameterSet/python/DummyCfis.py  modified in PR(s): #51413
  - File  FWCore/ParameterSet/python/Types.py  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/Entry.cc  modified in PR(s): #50585, #51413
  - File  FWCore/ParameterSet/src/ParameterDescription.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/ParameterDescriptionNode.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/ParameterSet.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/types.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc  modified in PR(s): #51413
  - File  FWCore/PythonParameterSet/src/PyBind11Module.h  modified in PR(s): #51413

cmsbuild commented:
Pull request #45216 was updated. @Dr15Jones, @cmsbuild, @makortel, @smuzaffar can you please check and sign again.

felicepantaleo commented:
@cmsbuild please test

felicepantaleo commented:
Ciao @makortel thanks for the detailed look. I have extended the tests you listed and, in doing so, found and fixed a real bug. 

- Infinity and NaN do propagate. they work today, for both `double` and `float`: `edm::encode` writes `inf`, `-inf`, `nan`, `edm::decode` restores them, and the full `ParameterSet::toString` round-trip preserves them. The reason the two lines were commented out is the test helper, not the framework: `do_compare` uses `CHECK(lhs == rhs)` and NaN never compares equal to itself, so the NaN case can never pass that way. I added a dedicated `testbodyNaN` / `testbodyVectorNaN` helper that checks `std::isnan` after the round-trip, and enabled the infinity cases for `double` as well as `float`.

- subnormals do not propagate: `std::stod` and `std::stof` throw `std::out_of_range` when the parsed result is subnormal, so `edm::decode` returns false and the job fails with

```
EntryError can not convert representation of x: 4.9406564584124654e-324 to value of type double
```

This is pre-existing behaviour that `float` inherits. `std::numeric_limits<T>::min()` (smallest normal), `max()` and `lowest()` all round-trip exactly, so I test those and document the subnormal limitation in a comment rather than asserting it.

- `std::numeric_limits<double>::min()`: your suspicion looks right
The `float` equivalent of that commented-out `REQUIRE` now passes in `ProducerWithPSetDesc` (`floatv1` defaults to `std::numeric_limits<float>::min()` and is checked for exact equality), which supports your reading that the "precision is 16 instead of 17". 
I left the `double` line as it is so as not to collide with what you are checking. I did not investigate the separate "cmsRun will fail with a value this big" comment on `vdoublev1`.

- the `ProducerWithPSetDesc` extension caught a bug, now fixed: adding the parameters there exposed that `float` was written into the generated cfi with only 6 significant digits, so it did not read back as the same value: `std::numeric_limits<float>::min()` came out as `1.17549e-38` instead of `1.17549435e-38`. The cause is that `writeSingleValue<float>` fell through to the generic template, which uses the default stream precision; `double` is correct only because of the explicit `formatDouble` plus its `writeSingleValue<double>` specialization. Fixed by mirroring that for `float` (`max_digits10`, then the shorter `digits10` form only when it converts back identically, so `0.3` still writes as `0.3`).

This is worth emphasising because the value round-trip tests in `test_catch2_ps.cc` all passed while this bug was present: `edm::encode`/`edm::decode` and the cfi writer are two independent paths, and only the integration test exercises the second one. 

`run_ParameterSet.sh` passes and so do the `FWCore/ParameterSet` unit tests.

I did not add the `edmProvDump` provenance test since you mentioned you were thinking of adding one. The float parameters in `ProducerWithPSetDesc` are now there to build it on.

Agreed that reading a `vector<double>` back as `vector<float>` would be a much smaller change. The argument for the distinct type is the one you make, that single and double precision configuration types mirror the existing int32/int64 signed and unsigned handling, and that the intent is visible in the configuration rather than only at the call site.

cmsbuild commented:
+1

**Size**: This PR adds an extra 80KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9ee9da/54925/summary.html
**COMMIT**: af69b627d71c767de41f633fb3a38fa3675e2127
**CMSSW**: CMSSW_20_1_X_2026-07-26-0000/el9_amd64_gcc13
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/45216/54925/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_20_1_X_2026-07-26-0000+9ee9da/77076/validateJR.html)
 - DQMHistoTests: Total files compared: 45
 - DQMHistoTests: Total histograms compared: 3559760
 - DQMHistoTests: [Total failures: 5](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_20_1_X_2026-07-26-0000+9ee9da/77076/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3559737
 - DQMHistoTests: Total skipped: 18
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 44 files compared)
 - Checked 195 log files, 163 edm output root files, 45 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_20_1_X_2026-07-26-0000+9ee9da/77076/triggerResults): no differences found

makortel commented:
> subnormals do not propagate: `std::stod` and `std::stof` throw `std::out_of_range` when the parsed result is subnormal, so `edm::decode` returns false and the job fails with
> 
> ```
> EntryError can not convert representation of x: 4.9406564584124654e-324 to value of type double
> ```
> 
> This is pre-existing behaviour that `float` inherits. `std::numeric_limits<T>::min()` (smallest normal), `max()` and `lowest()` all round-trip exactly, so I test those and document the subnormal limitation in a comment rather than asserting it.

I discovered this one as well. I think we can live with it, and maybe some day we get to convert the `types.cc` to use `std::from_chars()` that _might_ work better.

I also discovered the `cfi` generation does not work `inf` or `NaN` (the generated Python is not valid). I have a fix for this in my local branch for `double` that is straightforward to extend for `float` after this PR.

----

I found this script that should still be extended for `float`
https://github.com/cms-sw/cmssw/blob/d33e107ded45ad2ded12865f7844288961407030/FWCore/ParameterSet/scripts/edmPythonConfigToCppValidation#L33
https://github.com/cms-sw/cmssw/blob/d33e107ded45ad2ded12865f7844288961407030/FWCore/ParameterSet/scripts/edmPythonConfigToCppValidation#L51
(the script does not seem to have any tests)

----

This file needs to be extended
https://github.com/cms-sw/cmssw/blob/af69b627d71c767de41f633fb3a38fa3675e2127/FWCore/ParameterSet/src/FillDescriptionFromPSet.cc#L51-L52
I _guess_ a corresponding test could be in `ProducerWithPSetDesc` a `VPSet` whose default value contains a PSet with `float` or/(and) a `vfloat` element(s). I don't see any such systematic test in `ProducerWithPSetDesc`, so I don't ask for one in this PR.

----

This file needs to be extended
https://github.com/cms-sw/cmssw/blob/af69b627d71c767de41f633fb3a38fa3675e2127/FWCore/ParameterSet/src/ParameterDescriptionBase.cc#L171

It seems to me the corresponding test would be extending `ProducerWithPSetDesc` here
https://github.com/cms-sw/cmssw/blob/af69b627d71c767de41f633fb3a38fa3675e2127/FWCore/Integration/plugins/ProducerWithPSetDesc.cc#L1286-L1290
and here
https://github.com/cms-sw/cmssw/blob/af69b627d71c767de41f633fb3a38fa3675e2127/FWCore/Integration/plugins/ProducerWithPSetDesc.cc#L595

with `float` and `vfloat` parameter values (without default values around line 595, and with defaults around lines 1286-1290. There would be knock-on effects at least in `FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_cfi.py`. (Ideally the test would systematically test all parameter types, but I'm not asking for that for this PR).

felicepantaleo committed:
FWCore: handle float and vfloat in the remaining ParameterSet code paths

ParameterDescriptionBase::writeLabelValueCfi did not treat vfloat as a vector
type, so a vfloat default written into a generated cfi came out without the
enclosing brackets and the cfi was not valid Python. Also adds float and vfloat
to FillDescriptionFromPSet and to the edmPythonConfigToCppValidation script.

ProducerWithPSetDesc gains float and vfloat parameters without defaults, and a
vfloat set through the DescriptionCloner (to the same value as the description
default, as done for p_int) which is what exercises the writeLabelValueCfi path.
The briefdoc reference changes in many lines only because the added parameter
names widen the first column.

felicepantaleo committed:
FWCore/ParameterSet: decode floating point with std::from_chars

std::stod and std::stof report ERANGE for a subnormal result, so a subnormal
value such as std::numeric_limits<double>::denorm_min() could be encoded but not
read back. std::from_chars only reports result_out_of_range when the value is not
representable, so subnormals now round-trip; it is also locale independent and
does not allocate. Leading whitespace and a leading plus sign are stripped so the
accepted syntax stays the same, and trailing characters are still ignored.

The round-trip tests for denorm_min are enabled for both double and float.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45216/50329

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Integration/plugins/ProducerWithPSetDesc.cc  modified in PR(s): #51413
  - File  FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_briefdoc.txt  modified in PR(s): #51413
  - File  FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_cfi.py  modified in PR(s): #51413
  - File  FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_doc.txt  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/Entry.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/ParameterDescription.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/ParameterDescriptionNode.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/ParameterSet.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/types.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #51413
  - File  FWCore/ParameterSet/python/DummyCfis.py  modified in PR(s): #51413
  - File  FWCore/ParameterSet/python/Types.py  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/Entry.cc  modified in PR(s): #50585, #51413
  - File  FWCore/ParameterSet/src/ParameterDescription.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/ParameterDescriptionNode.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/ParameterSet.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/types.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/test/test_catch2_ps.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc  modified in PR(s): #51413
  - File  FWCore/PythonParameterSet/src/PyBind11Module.h  modified in PR(s): #51413

cmsbuild commented:
Pull request #45216 was updated. @Dr15Jones, @cmsbuild, @makortel, @smuzaffar can you please check and sign again.

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45216/50330

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Integration/plugins/ProducerWithPSetDesc.cc  modified in PR(s): #51413
  - File  FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_briefdoc.txt  modified in PR(s): #51413
  - File  FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_cfi.py  modified in PR(s): #51413
  - File  FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_doc.txt  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/Entry.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/ParameterDescription.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/ParameterDescriptionNode.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/ParameterSet.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/types.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #51413
  - File  FWCore/ParameterSet/python/DummyCfis.py  modified in PR(s): #51413
  - File  FWCore/ParameterSet/python/Types.py  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/Entry.cc  modified in PR(s): #50585, #51413
  - File  FWCore/ParameterSet/src/ParameterDescription.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/ParameterDescriptionNode.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/ParameterSet.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/types.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/test/test_catch2_ps.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc  modified in PR(s): #51413
  - File  FWCore/PythonParameterSet/src/PyBind11Module.h  modified in PR(s): #51413

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45216/50330/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45216/50330/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45216/50331

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Integration/plugins/ProducerWithPSetDesc.cc  modified in PR(s): #51413
  - File  FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_briefdoc.txt  modified in PR(s): #51413
  - File  FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_cfi.py  modified in PR(s): #51413
  - File  FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_doc.txt  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/Entry.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/ParameterDescription.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/ParameterDescriptionNode.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/ParameterSet.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/types.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #51413
  - File  FWCore/ParameterSet/python/DummyCfis.py  modified in PR(s): #51413
  - File  FWCore/ParameterSet/python/Types.py  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/Entry.cc  modified in PR(s): #50585, #51413
  - File  FWCore/ParameterSet/src/ParameterDescription.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/ParameterDescriptionNode.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/ParameterSet.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/types.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/test/test_catch2_ps.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc  modified in PR(s): #51413
  - File  FWCore/PythonParameterSet/src/PyBind11Module.h  modified in PR(s): #51413

cmsbuild commented:
Pull request #45216 was updated. @Dr15Jones, @cmsbuild, @makortel, @smuzaffar can you please check and sign again.

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 72KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9ee9da/54953/summary.html
**COMMIT**: 9f53fdd07b29a990689fdc01d0d0ebcd53945804
**CMSSW**: CMSSW_20_1_X_2026-07-27-2300/el9_amd64_gcc13
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/45216/54953/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_20_1_X_2026-07-27-2300+9ee9da/77097/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_20_1_X_2026-07-27-2300+9ee9da/77097/validateJR.html)
 - DQMHistoTests: Total files compared: 45
 - DQMHistoTests: Total histograms compared: 3588680
 - DQMHistoTests: [Total failures: 2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_20_1_X_2026-07-27-2300+9ee9da/77097/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3588660
 - DQMHistoTests: Total skipped: 18
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 44 files compared)
 - Checked 195 log files, 163 edm output root files, 45 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_20_1_X_2026-07-27-2300+9ee9da/77097/triggerResults): no differences found

felicepantaleo commented:
type ngt

makortel commented on FWCore/ParameterSet/src/types.cc:
```diff
@@ -428,6 +429,29 @@ bool edm::encode(std::string& to, std::vector<unsigned long long> const& from) {
+  // previous implementation reject values such as
+  // std::numeric_limits<double>::denorm_min(). std::from_chars only reports
+  // result_out_of_range when the value is not representable at all, so subnormals
+  // round-trip. It is also locale independent and does not allocate.
```
I don't see any need to refer to the earlier implementation at this level of detail. I'd be fine with something along "The std::from_chars handles subnormals better than std::stod/std::stof." so that it is clear for a future reader that the `std::from_chars` should not be replaced with `std::stod`/`std::stof`.

makortel commented on FWCore/ParameterSet/src/types.cc:
```diff
@@ -428,6 +429,29 @@ bool edm::encode(std::string& to, std::vector<unsigned long long> const& from) {
+    }
+    auto const* const first = from.data();
+    auto const* const last = first + from.size();
+    // As before, trailing characters after a valid number are ignored.
```
I'd prefer something along
```suggestion
    // Trailing characters after a valid number are ignored to be compatible with earlier behavior.
```

makortel commented on FWCore/ParameterSet/src/types.cc:
```diff
@@ -428,6 +429,29 @@ bool edm::encode(std::string& to, std::vector<unsigned long long> const& from) {
+  template <typename T>
+  bool decodeFloatingPoint(T& to, std::string_view from) {
+    // std::from_chars accepts neither leading whitespace nor a leading '+',
+    // both of which std::stod did, so strip them to keep the previous behaviour.
```
I'd like something along
```suggestion
    // so strip them to keep the earlier behaviour.
```
more

makortel commented on FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc:
```diff
@@ -0,0 +1,275 @@
+    std::set<edm::ParameterTypes> found;
+    using U = std::underlying_type_t<edm::ParameterTypes>;
+    for (int i = static_cast<int>(std::numeric_limits<U>::min()); i <= static_cast<int>(std::numeric_limits<U>::max());
+         ++i) {
```
Why not using `U` directly?

makortel replied:
Actually, after testing a few things to understand why the loop doesn't die on the `assert()` in `edm::parameterTypeEnumToString()`, I found that (on x86 EL8) the `U` is `unsigned int`, and `static_cast<int>(std::numeric_limits<U>::max())` leads to `-1`, and therefore the loop body gets never run.

makortel replied:
Another problem is the use of `<=` in the loop condition. Because the `++i` is executed first before checking the loop condition, when `i` is `max()` the condition passes, and after that the `++i` overflows. With `int` the overflow is undefined behavior, and with `unsigned int` `i` wraps back to zero, and the loop will never terminate.

makortel commented on FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc:
```diff
@@ -0,0 +1,275 @@
+      auto const type = static_cast<edm::ParameterTypes>(static_cast<U>(i));
+      if (not edm::parameterTypeEnumToString(type).empty()) {
+        found.insert(type);
+      }
```
I do not understand how this check can work. The `switch` in `edm::parameterTypeEnumToString()`'s has
https://github.com/cms-sw/cmssw/blob/9f53fdd07b29a990689fdc01d0d0ebcd53945804/FWCore/ParameterSet/src/ParameterDescriptionNode.cc#L104-L105
and we build the code without `-DNDEBUG` so the `assert()` should fire.

makortel replied:
The `assert()` could be worked around by moving the `switch` structure to a new function, e.g. `edm::parameterTypeEnumToStringUnchecked()` that returns an empty string as the `default` case of the `switch`. The `edm::parameterTypeEnumToString()` would then call `edm::parameterTypeEnumToStringUnchecked()` and `assert()` the return value to be non-empty. This test would call `edm::parameterTypeEnumToStringUnchecked()`.

makortel commented on FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc:
```diff
@@ -0,0 +1,275 @@
+      CHECK(stripped.front() == '[');
+      CHECK(stripped.back() == ']');
+    } else {
+      CHECK(stripped.front() != '[');
```
Shouldn't this case also `REQUIRE(not stripped.empty());' (after which it could be moved outside of the `if`-`else` blocks)?

makortel commented on FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc:
```diff
@@ -0,0 +1,275 @@
+  }
+
+  template <typename T>
+  void checkType(edm::ParameterTypes expected, bool isVector, T const& value) {
```
Looks like the `isVector` could be deduced from the `T`.

makortel commented on FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc:
```diff
@@ -0,0 +1,275 @@
+// So rather than testing one type, these tests enumerate the types themselves:
+// the set of types the framework names is discovered by scanning the enum's
+// underlying value range, and every type found MUST be covered here. A new type
+// therefore fails this test until it is added to the table below.
```
Could you edit this description of the tests more concise?

makortel commented on FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc:
```diff
@@ -0,0 +1,275 @@
+              std::set<edm::ParameterTypes>& covered,
+              std::vector<std::string>& names) {
+    ps.addParameter<T>(name, value);
+    covered.insert(type);
```
Looks like the `type` parameter could be removed with
```suggestion
    covered.insert(edm::ParameterTypeToEnum::toEnum<T>());
```
plus special cases for `T` being `edm::ParameterSet` or `std::vector<edm::ParameterSet>`.

makortel commented on FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc:
```diff
@@ -0,0 +1,275 @@
+}  // namespace
+
+TEST_CASE("every parameter type is handled consistently", "[ParameterTypes]") {
+  testedTypes.clear();
```
I'd prefer to have the `testedTypes` defined locally inside the relevant `SECTION` (instead of using global variables) and checked there. The VPSet `SECTION` already follows that pattern.

makortel commented on FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc:
```diff
@@ -0,0 +1,275 @@
+    // fillDescriptionFromPSet, which looks each parameter up in a per-type table and
+    // SILENTLY SKIPS anything missing from it. A type absent from that table is
+    // therefore dropped from the generated cfi without any error, which is what
+    // happened to float and vfloat.
```
Could you make this comment more concise? I think it would be better to not refer to present implementation of `fillDescriptionFromPSet()`, and focus on the required behavior instead.

makortel commented on FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc:
```diff
@@ -0,0 +1,275 @@
+    for (auto type : namedParameterTypes()) {
+      if (knownUnsupported.find(type) != knownUnsupported.end()) {
+        continue;
+      }
```
I think it would be better to deal with the `vbool` special case already in the `namedParameterTypes()`.

makortel commented on FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc:
```diff
@@ -0,0 +1,275 @@
+    // k_vbool is named by the framework but cannot be used as a parameter: edm::Entry
+    // has the kTvBool code with no std::vector<bool> constructor or getter, so
+    // ParameterDescription<std::vector<bool>> does not compile. Left as is here; this
+    // records the gap rather than hiding it.
```
Please make the comment more concise (if kept after addressing other comments).

makortel reviewed (commented):
I appreciate the new test and making the `decode()` to work with subnormals, but want to note that substantial code additions or changes impact the review time as well.

felicepantaleo committed:
FWCore/ParameterSet: test every parameter type systematically

One list of sample values drives both checks: each type maps to its enum and
writes a well formed untyped cfi, and every type in a VPSet default element
survives into the written cfi.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45216/50354

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Integration/plugins/ProducerWithPSetDesc.cc  modified in PR(s): #51413
  - File  FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_briefdoc.txt  modified in PR(s): #51413
  - File  FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_cfi.py  modified in PR(s): #51413
  - File  FWCore/Integration/test/unit_test_outputs/testProducerWithPsetDesc_doc.txt  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/Entry.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/ParameterDescription.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/ParameterDescriptionNode.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/ParameterSet.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/interface/types.h  modified in PR(s): #51413
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #51413
  - File  FWCore/ParameterSet/python/DummyCfis.py  modified in PR(s): #51413
  - File  FWCore/ParameterSet/python/Types.py  modified in PR(s): #51413
  - File  FWCore/ParameterSet/scripts/edmPythonConfigToCppValidation  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/Entry.cc  modified in PR(s): #50585, #51413
  - File  FWCore/ParameterSet/src/FillDescriptionFromPSet.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/ParameterDescription.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/ParameterDescriptionBase.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/ParameterDescriptionNode.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/ParameterSet.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/src/types.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/test/test_catch2_allParameterTypes.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/test/test_catch2_ps.cc  modified in PR(s): #51413
  - File  FWCore/ParameterSet/test/test_catch_ParameterSetDescription.cc  modified in PR(s): #51413
  - File  FWCore/PythonParameterSet/src/PyBind11Module.h  modified in PR(s): #51413

cmsbuild commented:
Pull request #45216 was updated. @Dr15Jones, @cmsbuild, @makortel, @smuzaffar can you please check and sign again.

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 56KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9ee9da/54978/summary.html
**COMMIT**: 98976752ac57e575c1a85597e602da966c97a1f2
**CMSSW**: CMSSW_20_1_X_2026-07-28-2300/el9_amd64_gcc13
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/45216/54978/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_20_1_X_2026-07-28-2300+9ee9da/77139/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_20_1_X_2026-07-28-2300+9ee9da/77139/validateJR.html)
 - DQMHistoTests: Total files compared: 45
 - DQMHistoTests: Total histograms compared: 3588680
 - DQMHistoTests: [Total failures: 9](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_20_1_X_2026-07-28-2300+9ee9da/77139/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3588653
 - DQMHistoTests: Total skipped: 18
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 44 files compared)
 - Checked 195 log files, 163 edm output root files, 45 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_20_1_X_2026-07-28-2300+9ee9da/77139/triggerResults): no differences found

makortel commented:
Comparison differences are related to https://github.com/cms-sw/cmssw/issues/47071

makortel commented:
+core

Thanks for the contribution. @felicepantaleo Could you add to the PR description that this PR also replaces to used of `std::stod` with `std::from_chars` in order to deal better with subnormals?

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @ftenchini, @sextonkennedy, @mandrenguyen (and backports should be raised in the release meeting by the corresponding L2)

mandrenguyen commented:
+1
